### ISSUE TITLE
Always use table.insert() when changing vanilla prototypes.

### ISSUE THREAD

**Issue by lossycrypt** [Factorio-Xander-Mod/Xander-Mod#2]
In misc.lua you're hardcoding the crafting_categories for the base player prototype. This is fundamentally wrong and if any two mods that do this are ever loaded at the same time it will invisibly break for the mod that is earlier in the loading order. You MUST ALWAYS use table.insert when changing that kind of thing.

table.insert(data.raw.player["player"].crafting_categories,"basic-sluice")

ps: Welcome to github.

**Comment by lossycrypt** [comment]
pps: I'm just stating this very explicitly because that kind of lingering incompatibility makes me cringe. Especially because it would break my own mods if someone used them together with yours.

**Comment by lossycrypt** [comment]
ppps: Browsing a bit more i see you're doing this to the chemplant too (prototypes/entity/production/chemical.lua) and probably in a few other places. If your intention is to only remove "chemistry" from the vanilla chemplant you should use table.remove() if you want to move all crafting categories to your new assembling machine (checmical reactor?) then you need to change the crafting categories during data-updates.lua to make sure you can see any new categories added by other mods.

**Comment by ironchefpython** [comment]
I totally understand what you're saying from an extensiblilty standpoint, but does it really apply if a mod (like this one) as replaced every single vanilla item and recipe and research in the base game? I'm pretty sure as it stands Xander isn't compatible with any mod that depends on vanilla recipes, technologies, ores, or items... so that leaves.... https://mods.factorio.com/mods/GotLag/Honk

I understand your point about best practices, but do you still think it's necessary to use table.insert when the desired goal is to blow away all other categories and vanilla data?

**Comment by Repofme1** [comment]
Wow, a lot of stuff that's half over my head!  I generally see what's going on; I actually used that [x].variable = {} a lot because it's the only way I knew how to do it...  And yeah, in some cases I did it to remove categories, etc. from vanilla things just because the default ones didn't fit the naming scheme.  This whole thing would really fall apart without somewhat memorable names.

I started this mod almost exactly a year ago with a few small things, just to see if I could get it to work, after a slow and frustrating attempt to mod Minecraft several years back.  Then over the summer I exploded it with the huge amount of stuff it has now, but still sort of intending it as a personal project, until one day I realized, ...wait this is big enough to publish.  In short, I don't actually properly know Lua, or the Factorio API, just what I can pick up from other mods :/  Like data-updates.lua, for example, I don't _really_ know what that does.  And I didn't even figure out migrations until after I published the mod.  So I apologize if my poor coding breaks in contact with your mods, I still have a lot to learn.  I certainly could use table.insert(), but I'd just have to use table.remove() too (if there is one, I figure); this might be a more robust way to do it overall.

My interests and knowledge are mostly in the materials and industrial chemistry (hence the content of the mod), and I sort of intend for the XM materials and parts to be a very realistic basis for making things.  The only reason I re-use as much of vanilla as I do is for the sake of saving some effort and code.  I guess like ironchef said, mods using much of vanilla may not even make logical sense alongside Xander Mod, even on a conceptual basis.

**Comment by lossycrypt** [comment]
TL;DR:
I suggest this soley because i think this is a pure programming "bug" that when fixed as no negative downsides to either yours or other mods. If i had a better understanding of your mod internals i might have even done a pull request.

############

I put some further thought into it an i think crafting-categories are really the main thing you can do wrong here. Changing recipes of vanilla items is fine as long as you don't make them unobtainable (which you don't). Putting custom crafting categories on machines that are created by your own mod is fine too ofc.

############

For starters my argument is purely technical for. Discussion if other mods running "on top of" xanders still look and feel logical/realistic can be done at a later point. I haven't been able to get to endgame in xanders yet, but from what i've seen so far i see no reason to sacrifice compatibility, or any reason why compatibility shouldn't be possible.

@ironchefpython:
I don't think the "desired goal" here was to purposefully break vanilla. As Repofme1 stated his goal was realism.
From a purely technical standpoint all vanilla items still exist, if any mod implements a recipe with "iron-gear-wheel" or "electronic-circuit" it will still work, even though the naming/graphics/means of production for those components will have changed. So imho any mod that does not rely on crafting_categories should still be able to run on top of xanders even now.

@repofme1:
You stated that you leared from reading other mods. Which we all did i think. And i'm pretty sure that i also did overwrite categories back when i didn't know better. But now that i do i think i am obliged to correct this mistake in my own mods, so that "the next generation of modders" - so to speak - does it right in the first place when they learn from my mods. And i'm just asking you to do the same.

But for example if i invented a new category "fancy-crafting" that my mod relies on and that can be hand-crafted or made in standard assemblers OR in late game using fancy-assemblers then your overwriting the categories makes it essentially impossible to make my stuff until the very late game, and the worst thing about it is that there will be no error messages about this. Players that have never played my mod wouldn't even notice that there missing something, or why they're missing it.

One of my mods also requires a new hand-crafting category for boostrapping. If it's overwritten then progression in that mod becomes impossible, but the player might only notice this after spending tens of hours on a map.

So, to fix this my suggestion is to make all changes to vanilla crafting categories in one go in data-updates.lua. As this only affects 7 vanilla-prototypes i think (player, assmebler 1-3, furnace, chemplant, refinery?) it should be fairly short and straight-forward. You can use table.insert(table,value) to insert new categories and table.remove(table,value) to remove existing categories. You can also use local new = table.deepcopy(table) to copy a table, for example to copy all categories (including other-mod added ones) from the vanilla-chem-plant to one of your higher tier chem plants and then overwrite the categories on the vanilla chem-plant as usual, this way other mods could even benefit from your higher-tier machines.

data-updates.lua is just the file that is executed after data.lua btw, you can read this up on:
http://lua-api.factorio.com/latest/Data-Lifecycle.html

**Comment by Repofme1** [comment]
Okay, I think it's more clear to me now.  I'll re-write the crafting category changes to use table.add(), that shouldn't be too much work come to think about it.

Out of curiosity, would you mind telling me which mod of yours it is that has hand crafting?  I see bootstrapping and it piques my interest as potentially very cool : )

**Comment by lossycrypt** [comment]
It's still in development and i'm kindq reluctant to talk about anything before i have at least an alpha release. It's probably not the kind of boostraping you imagine though. I intend the mod to affect base gameplay as little as possible. So that even if it's installed the player can choose to completely ignore it (not sure why they'd install it then :P).

Also btw it's table.insert() not table.add()

**Comment by Repofme1** [comment]
Alright, latest update to XM 1.4.0 should resolve this.  Let me know if problems continue.
